### PR TITLE
Refactoring report generation module a bit

### PR DESCRIPTION
**Pull request by baev**
### NOT READY TO MERGE YET

Generate report using [ModelMapper](http://modelmapper.org/) and Groovy instead XSLT

## Features 

* New plugin system 
  * *Enrich plugins* - you can preprocess test results. An example load some info about tests from TMS
  * *Tab plugins* - each report tab it's plugin now. You can add your own tabs to report
  * *Widget plugins* - overview now contains widgets - each tab plugin can provide widget
* Group suites by name (from different files)

## Not implemented yet

* Migrations (we going to change migrations logic in separate PR)
* Attachments copying (use old provider, will be removed soon)
* Environment (use old provider, will be removed soon)
* Plugins in report face 

## Problems

* Ignore start/stop times for test suites. It will be fixed in next releases (most probably in 2.0)

## Increase performance
**4 times faster, 10 times less memory usage**

Before:

After:

### CODE REVIEW

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/XUnit.groovy [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+class XUnit implements TabPlugin {
+
+    AllureXUnit xUnit = new AllureXUnit(time: new Time(start: Long.MAX_VALUE, stop: Long.MIN_VALUE));
+
+    private Map<String, AllureTestSuite> testSuites = new HashMap<>();
+
+    @Override
+    void process(AllureTestCase testCase) {
+        if (!testCase.status) {
+            throw new ReportGenerationException("Test case status should not be null")
```
Comment (lanwen) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/XUnit.groovy]: what the link between status and null-message?

**Comment by lanwen** [comment]
a bit. :100:

File: allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/XUnit.groovy (hunk 1 above)
Comment (baev) [inline allure-report-data/src/main/groovy/ru/yandex/qatools/allure/data/plugins/XUnit.groovy]: ?

**Comment by baev** [comment]
@vania-pooh done. Thanks!

**Comment by baev** [comment]
We going to merge this PR, release 1.4.7, and start huge refactoring of plugin system

**Comment by pupssman** [comment]
Wow, looks great!
What about that fancy "plugin system"? Is it possible to contribute them at some point?

**Comment by pupssman** [comment]
``Group suites by name (from different files)`` -- can you give more info?

**Comment by vania-pooh** [comment]
@pupssman Allure 1.4.7 release main purpose was to prepare for using plugins without changing report itself. During the next iterations @baev and @eroshenkoam will add more plugin-related stuff and will also change XML.

**Comment by baev** [comment]
> What about that fancy "plugin system"

It's first step to new plugin system. Using plugins you can customize report (add new tabs, widgets etc). Will be available since Allure 2.0

> Group suites by name (from different files) -- can you give more info?

Test suites with the same `name` will be merged into one in the report.

**Comment by pupssman** [comment]
Is it released, guys?

**Comment by baev** [comment]
@pupssman yep

https://github.com/allure-framework/allure-core/releases

**Comment by pupssman** [comment]
That thing says `pre-release`. Is it on central?

**Comment by baev** [comment]
@pupssman yep

**Comment by pupssman** [comment]
It is strange, but `1.4.7` and `1.4.5` reports built from same data have different run time:
5_time
7_time

**Comment by baev** [comment]
@pupssman seems like different rounding

**Comment by pupssman** [comment]
@baev indeed so. But why? Whill that difference build up for many tests?

**Comment by baev** [comment]
> Whill that difference build up for many tests?

Nope. I guess it's the same number rounded up in 1.4.5, and rounded down in 1.4.7. So, you can check it - just compare run time at `xunit.json`

**Pull request merged**